loki/frontend/source.py

```diff
@@ -127,6 +156,12 @@ def clone_lines(self, span=None):
             for idx, line in enumerate(self.string.splitlines())
         ]
 
+    def invalidate(self):
+        """
+        Return an invalid copy of this source with ``status=SourceStatus=INVALID_NODE``.
+        """
+        return self.clone(status=SourceStatus.INVALID_NODE)
+
```

Comment:
This seems only to be used from tests, not in the actual transformer. Is that intentional?

Reply:
Sort of. It's a neat utility that can freely be used by user-level transformations, when for example a changed clone of a node is used in a dict-based `Transformer` replacement (as in the tests). However, the only core-level automatic invalidation that happens in this PR is based by in-place changes to the status, so it does not use the utility.